File: app/src/main/java/com/example/androidkotlincompose/modules/home/posts/data/source/remote/PostsRemoteDataSource.kt
package com.example.androidkotlincompose.modules.home.posts.data.source.remote

import com.example.androidkotlincompose.apiservices.dummyjson.DummyJsonApiService
import com.example.androidkotlincompose.core.models.NetworkResult
import com.example.androidkotlincompose.modules.home.posts.data.response.PostsResponse
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext

class PostsRemoteDataSource(
    private val apiService: DummyJsonApiService,
    private val dispatcher: CoroutineDispatcher
) {
    suspend fun getPosts(): NetworkResult<PostsResponse> {
        return withContext(dispatcher) {
            try {
                val response = apiService.getPosts()
                if (response.isSuccessful) {
                    val body = response.body()
                    if (body == null) {
                        NetworkResult.Error("Response body was null")
                    } else if (body.posts.isNullOrEmpty()) {
                        NetworkResult.Empty
                    } else {
                        NetworkResult.Success(body)
                    }
                } else {
                    NetworkResult.Error(response.message())
                }
            } catch(e: Exception) {
                NetworkResult.Error(e.message ?: "Something went wrong!")
            }
        }
    }
}

File: mobile_flutter/lib/data/remote_data_source/posts/posts.remote_data_source.dart
import 'package:dio/dio.dart';
import 'package:mobile_flutter/core/common_models/network.model.dart';
import 'package:mobile_flutter/data/remote_data_source/posts/models/posts_response.model.dart';

class PostsRemoteDataSource {
  final Dio _dio;

  PostsRemoteDataSource(this._dio);

  Future<NetworkResult<PostsResponse>> getPosts() async {
    try {
      final response = await _dio.get('posts');
      final data = PostsResponse.fromJson(response.data);

      if (data.posts.isEmpty) {
        return NetworkResult.empty();
      }

      return NetworkResult.success(data);
    } catch (e) {
      return NetworkResult.error(e.toString());
    }
  }
}
